Regression: who/utmp only shows the last XRDP user after display-string migration (ut_id collisions)

### Summary

After recent display-string changes, `who` on the server only shows the most recently logged-in XRDP user, instead of all concurrent XRDP sessions.

This appears to be a regression in `sesman/sesexec/sessionrecord.c` utmp ID generation.

---

### Environment

- xrdp branch: `devel`
- Reproduced after updates including:
  - `656a125c` (`utmp: Remove display number from interface`)
  - display string changes merged via PR #3729
- Linux with utmp enabled (`who` output affected)

---

### Previous behavior (before update)

Multiple concurrent XRDP users were visible in `who`, e.g.:

- `xrdp:60`
- `xrdp:61`
- `xrdp:62`

---

### Current behavior (after update)

Only one XRDP session (typically the last one) remains visible in `who`, e.g.:

- `xrdp:X11-88`

Even with multiple users connected, earlier entries are overwritten.

---

### Steps to reproduce

1. Start xrdp/sesman with current `devel`.
2. Connect multiple concurrent XRDP users.
3. Run `who` on server.
4. Observe only one XRDP user line remains (or lines overwrite each other).

---

### Root-cause analysis

In `sesman/sesexec/sessionrecord.c`, `get_display_num()` derives a numeric ID from display string for `ut_id`.

Current logic extracts the first digit run in the display string.
For displays like `X11-88`, `X11-89`, `X11-90`, this yields `11` for all sessions.

So all sessions get the same `ut_id` (e.g. `:%x` => `:b`), and `pututxline()` overwrites previous entries.

---

### Impact

- In multi-user scenarios, server-side tools can no longer reliably obtain the full logged-in user list.
- Session-name dependent applications can break (example: IME/input-method integrations that rely on session identity).
- Incorrect operational visibility (`who`, utmp-based auditing/monitoring).

---

### Suggested fix

For display strings like `x11-<num>` / `wayland-<num>`, derive utmp ID from the numeric part after `-` to avoid collisions.

(Alternative acceptable approach: any stable collision-resistant mapping of display string to `ut_id` constrained by field size.)

build failure:
~~~
  CC       eicp_process.o
eicp_process.c: In function ‘process_create_session_response’:
eicp_process.c:133:13: error: implicit declaration of function ‘strlcpy’; did you mean ‘strncpy’? [-Werror=implicit-function-declaration]
             strlcpy(s_item->display, display, sizeof(s_item->display));
             ^~~~~~~
             strncpy
cc1: all warnings being treated as errors
~~~

The build failure is introduced by commit c4727ad8 (Replace X11 display number with a display string).

  In sesman/eicp_process.c, this commit changed the display handling from int to const char * and added:

  strlcpy(s_item->display, display, sizeof(s_item->display));

  However, the file does not include string_calls.h, which declares strlcpy() in this project.
  With -Werror=implicit-function-declaration, this becomes a hard error:

  - implicit declaration of function 'strlcpy'
  - build stops at eicp_process.o

  So the root cause is: strlcpy() was introduced in eicp_process.c without adding its required header include.

--- a/sesman/eicp_process.c
+++ b/sesman/eicp_process.c
@@ -40,6 +40,7 @@
 #include "sesman.h"
 #include "sesman_access.h"
 #include "sesman_config.h"
+#include "string_calls.h"
 #include "guid.h"
 
 /******************************************************************************/

--- a/sesman/libsesman/verify_user_pam.c
+++ b/sesman/libsesman/verify_user_pam.c
@@ -400,9 +400,20 @@ auth_uds(const char *user, enum scp_login_status *errorcode)
 static int
 auth_start_session_private(struct auth_info *auth_info, const char *display)
 {
-    int error;
+    // For Linux and maybe other systems, pam_systemd needs to know the
+    // session type in order to set the session up correctly
+    const char *session_type;
+    if (g_get_x11_display_from_display_string(display) >= 0)
+    {
+        session_type = "x11";
+    }
+    else
+    {
+        session_type = "wayland";
+    }
+    g_setenv_log("XDG_SESSION_TYPE", session_type, 1);
 
-    error = pam_set_item(auth_info->ph, PAM_TTY, display);
+    int error = pam_set_item(auth_info->ph, PAM_TTY, display);
 
     if (error != PAM_SUCCESS)
     {

--- a/sesman/libsesman/verify_user_pam_userpass.c
+++ b/sesman/libsesman/verify_user_pam_userpass.c
@@ -209,9 +209,20 @@ auth_uds(const char *user, enum scp_login_status *errorcode)
 static int
 auth_start_session_private(struct auth_info *auth_info, const char *display)
 {
-    int error;
+    // For Linux and maybe other systems, pam_systemd needs to know the
+    // session type in order to set the session up correctly
+    const char *session_type;
+    if (g_get_x11_display_from_display_string(display) >= 0)
+    {
+        session_type = "x11";
+    }
+    else
+    {
+        session_type = "wayland";
+    }
+    g_setenv_log("XDG_SESSION_TYPE", session_type, 1);
 
-    error = pam_set_item(auth_info->ph, PAM_TTY, display);
+    int error = pam_set_item(auth_info->ph, PAM_TTY, display);
 
     if (error != PAM_SUCCESS)
     {

--- a/sesman/sesexec/sessionrecord.c
+++ b/sesman/sesexec/sessionrecord.c
@@ -71,9 +71,9 @@ typedef struct utmp _utmp;
 #include "string_calls.h"
 
 #define XRDP_LINE_FORMAT "xrdp:%s"
-// ut_id is a very small field on some platforms, so use the display
-// number in hex
-#define XRDP_ID_FORMAT ":%x"
+// ut_id is a very small field on some platforms, so use a discriminator
+// (e.g. ':' for x11) and a number in hex
+#define XRDP_ID_FORMAT "%c%x"
 
 /******************************************************************************/
 /**
@@ -106,30 +106,46 @@ str2memcpy(void *dest, const char *src, size_t n)
 
 /******************************************************************************/
 /**
- * Get a display number from the display string
+ * Set the idbuff string
  *
- * Converts the display string to some sort of number to write to the
- * extremely short ut_id field
- * @param display string
- * @return display number
+ * The ut_id field is extremely short (e.g. 4 usable characters on
+ * Linux/FreeBSD).
+ *
+ * @param buff Output buffer, including terminator
+ * @param bufflen Length of above
+ * @param display Display string for the session
  */
-static unsigned int
-get_display_num(const char *display)
+static void
+set_idbuff_str(char buff[], unsigned int bufflen, const char *display)
 {
-    unsigned int result = 0;
-    /* Look for the first digit in the string */
-    while (*display != '\0' && !isdigit(*display))
-    {
-        ++display;
-    }
-    /* convert consecutive digits to a number */
-    while (*display != '\0' && isdigit(*display))
-    {
-        result = (result * 10) + (*display - '0');
-        ++display;
-    }
+    unsigned int display_num = 0;
+    char discriminator;
 
-    return result;
+    // X11 display?
+    int x11_display = g_get_x11_display_from_display_string(display);
+    if (x11_display >= 0)
+    {
+        discriminator = ':';
+        display_num = (unsigned int)x11_display;
+    }
+    else
+    {
+        discriminator = 'W';
+        /* Look for the first digit in the string */
+        while (*display != '\0' && !isdigit(*display))
+        {
+            ++display;
+        }
+        /* convert consecutive digits to a number */
+        display_num = 0;
+        while (*display != '\0' && isdigit(*display))
+        {
+            display_num = (display_num * 10) + (*display - '0');
+            ++display;
+        }
+    }
+    display_num = MIN(display_num, 0xfff);
+    g_snprintf(buff, bufflen, XRDP_ID_FORMAT, discriminator, display_num);
 }
 
 /******************************************************************************/
@@ -158,11 +174,8 @@ add_xtmp_entry(int pid, const char *display,
 
     struct timeval tv;
 
-    /* Convert the display string to a number */
-    unsigned int display_num = get_display_num(display);
-
     g_memset(&ut, 0, sizeof(ut));
-    g_snprintf(idbuff, sizeof(idbuff), XRDP_ID_FORMAT, display_num);
+    set_idbuff_str(idbuff, sizeof(idbuff), display);
     g_snprintf(linebuff, sizeof(linebuff), XRDP_LINE_FORMAT, display);
     gettimeofday(&tv, NULL);
 

